pkg/apis/crd/v1alpha2/types.go

```diff
@@ -543,3 +543,88 @@ type TrafficControlList struct {
 
 	Items []TrafficControl `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BGPPolicy defines BGP configuration applied to Nodes.
+type BGPPolicy struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard metadata of the object.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec BGPPolicySpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type BGPPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []BGPPolicy `json:"items"`
+}
+
+// BGPPolicySpec defines the specification for a BGP policy.
+type BGPPolicySpec struct {
+	// NodeSelector is utilized to select Nodes to which the policy is applied.
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector"`
+
+	// LocalASN is the AS number used by BGP protocol.
+	LocalASN int32 `json:"localASN"`
+
+	// ListenPort is the port on which BGP protocol listen.
+	ListenPort *int32 `json:"listenPort,omitempty"`
+
+	// Advertisements selects IPs to be advertised to external BGP peers.
+	Advertisements Advertisements `json:"advertisements"`
+
+	// BGPPeers are the list of external BGP peers.
+	BGPPeers []BGPPeer `json:"bgpPeers"`
+}
+
+type Advertisements struct {
+	ServiceIPs *ServiceAdvertisement `json:"serviceIPs,omitempty"`
+
+	PodIPs *PodAdvertisement `json:"podIPs,omitempty"`
+
+	// Determine whether to advertise local Egress IPs.
+	EgressIPs bool `json:"egressIPs,omitempty"`
+}
+
+type ServiceAdvertisement struct {
+	// Determine whether to advertise ClusterIPs.
+	ClusterIPs bool `json:"clusterIPs,omitempty"`
+
+	// Determine whether to advertise ExternalIPs.
+	ExternalIPs bool `json:"externalIPs,omitempty"`
+
+	// Determine whether to advertise LoadBalancerIPs.
+	LoadBalancerIPs bool `json:"loadBalancerIPs,omitempty"`
+}
+
+type PodAdvertisement struct {
+	// Determine whether to advertise local Pod CIDR.
+	Enable bool `json:"enable,omitempty"`
+}
+
+type BGPPeer struct {
+	// Address is the IP address of the external BGP peer.
+	Address string `json:"address"`
+
+	// Port is the port number of the external BGP peer.
+	Port *int32 `json:"port,omitempty"`
+
+	// ASN is the ASN of the external BGP peer.
+	ASN int32 `json:"asn"`
+
+	// AuthSecretRef is a reference to a Secret storing the authentication password of the external BGP peer.
+	AuthSecretRef *string `json:"authSecretRef,omitempty"`
```

Comment:
How does it work exactly? In which Namespace are the secrets defined? I assume they need to be created in the same Namespace as Antrea. Does that mean Antrea will need permissions to read all secrets in the kube-system Namespace (I don't think this is the case today).

Reply:
This is a thoughtful question. I found the corresponding implementation in other CNIs:

1. Users assign a Secret reference in a CR, then create the corresponding Serect to store the password. To read the password, the CRD controller needs the `get` permission to read the corresponding Secret in the kube-system Namespace.
2. Uses set the base64 coded password in an annotation. 

I think we could let users set the based64 coded password in a BGPPolicy CR directly. Though it is not a good practice, it is easy for users to use. 

If users want to secure their BGP passwords, they could add a password for a BGP peer to the Serect named `antrea-bgp-passwords` in the kube-system Namespace (assuming that Antrea is installed in the Namespace). 

The secret might be like this:

```
apiVersion: v1
kind: Secret
metadata:
  name: antrea-bgp-passwords
  namespace: kube-system
type: Opaque
data:
  192.168.1.1-65512: dmVyeS1zZWNyZXQ=
  192.168.1.2-65513: dmVyeS1zZWNyZXQ=
```

Antrea needs the permission `get`, maybe `watch` (to watch the update of the secret)

```
  - apiGroups:
      - ""
    resources:
      - secrets
    resourceNames:
      - antrea-bgp-passwords
    verbs:
      - get
      - watch
```

Reply:
I think we should just support the Secret-based method.
The idea of having a single Secret, with a well-known name, mapping BGP peer IDs to passwords, is fine by me. We can also provide a configuration parameter for users to change the Secret name if they want to.